---
title: "coex"
subtitle: "An R package for straightforward co-expression analyses"
author: "Sam Buckberry"
output: BiocStyle::html_document
#output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{coex}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

# Getting started with `coex`

## Install and load required packages

Load the packages required for this vignette.
```{r setup, message=FALSE}
library(coex)
library(recount)
library(magrittr)
library(stringr)
library(cowplot)
```

## Download RNA-seq counts data
Load data from SRP041675 as a RangedSummarizedExperiment. This loads an object named `rse_gene`.
```{r, eval=FALSE}
file_url <- recount::download_study('SRP041675', type = "rse-gene",
                           download = FALSE)

# Load the data
load(file = url(file_url))
rse_gene
```


```{r}
load("~/Desktop/rse-gene.Rdata")
```

# Prepare the `CoexList` object 

## Load counts data into CoexList
Load the gene expression counts into a CoexList object
```{r}
cl <- CoexList(counts = assay(rse_gene))
```

### Load metadata into CoexList 
Add the gene and sample metadata to the CoexList object
```{r}
colData(cl) <- colData(rse_gene)
rowData(cl) <- rowData(rse_gene)
```

Reformat some of the metadata for analysis purposes
```{r}
group_df <- colData(cl)$title %>% 
    stringr::str_split(pattern = "_", n = 3, simplify = TRUE) %>%
    DataFrame()
group_df$V3[group_df$V2 == "Fresh"] <- "None"
group_df$V1 <- colnames(cl)
colnames(group_df) <- c("Sample", "Timepoint", "Treatment")

colData(cl) <- group_df[ ,c(2:3)]
rownames(colData(cl)) <- group_df[ ,1]
```

Now we can remove the `rse_gene` object
```{r}
rm(rse_gene)
```

# Pre-analysis filtering

```{r}
hist(edgeR::cpm(assay(cl), log=TRUE))
```

```{r}
keep <- edgeR::filterByExpr(assay(cl))
hist(edgeR::cpm(assay(cl)[keep, ], log=TRUE))

cl <- cl[keep, ]
```


# Normalisation

The goal of within sample normalisation is...

Count per million normalisation
```{r}
cl <- normCounts(cl, normMethod = "CPM")
```

CTF normalisation as described in... 

Normalise the data for co-expression calculations using the CTF method and scale using `asinh()`.
https://github.com/krishnanlab/RNAseq_coexpression/blob/main/src/CTF_normalize.R
```{r}
cl2 <- normCounts(cl, normMethod = "CTF", scaleMethod = "asinh")
```


# Clustering for quality control

First, lets do some plots to inspect for any batch effects or outlier samples

PCA plot
```{r, eval=FALSE}
pca_cpm_treat <- plotPCA(cl, groupVar = "Treatment") + ggtitle("CPM")
pca_ctf_treat <- plotPCA(cl2, groupVar = "Treatment") + ggtitle("CTF")

pca_cpm_time <- plotPCA(cl, groupVar = "Timepoint") + ggtitle("CPM")
pca_ctf_time <- plotPCA(cl2, groupVar = "Timepoint") + ggtitle("CTF")

cowplot::plot_grid(pca_cpm_treat, pca_ctf_treat,
                   pca_cpm_time, pca_ctf_time)
```

```{r, eval = FALSE}
pheatmap::pheatmap(cl@normCounts, kmeans_k = 500, show_rownames = FALSE,
                         scale = 'row', cluster_rows = TRUE,fontsize_col = 2,
                         annotation_col = as.data.frame(colData(cl)))

```

# Variance filtering

Filter the data
```{r, eval=FALSE}
checkFilter(cl)
cl <- applyFilter(cl, propGenes = 1)
```

# WGCNA analysis

## Soft-power calculations
Calculate soft power thresholds for WGCNA network
```{r, eval=FALSE}
cl <- calcSoftPower(cl)
```

Check what the estimated power is
```{r, eval=FALSE}
cl@powerEstimate
```

Plot soft power estimates
```{r, eval=FALSE}
plotSoftPower(cl)
```

## Adjacency matrix
Calc adjacency matrix
```{r, eval=FALSE}
cl <- calcAdjacency(cl, method = "wgcna", softPower = 5)
```

## Topological overlap matrix
Calculate topological overlap matrix (TOM)
```{r, eval=FALSE}
cl <- calcTOM(cl)
```

## Gene-wise clustering
Cluster genes
```{r, eval=FALSE}
cl <- calcTree(cl)
```

Plot clustering
```{r, eval=FALSE}
plotModuleTree(cl, pamStage=TRUE)
```

## Module eigengenes

Calculate module eigengenes for chosen tree split
```{r, eval=FALSE}
cl <- calcModuleEigengenes(cl, deepSplit = 3, minClusterSize = 30)
```

Calculate module membership. 
```{r, eval=FALSE}
cl <- calcModuleMembership(cl)
```




---

### Unused functions
```{r, eval=FALSE}
#' Plot dendrogram of samples from normalised data with colourbar
#'
#' @param cl An object of class CoexList.
#' @param plotColVar logical. Plot colorbar of colData variable defined in colVar.
#' @param colVar Character or numeric that refers to a
#' column name of colData(cl)
#' @param corMethod Correlation method. See ?cor
#' @param hclustMethod clustering method. See ?hclust
#' @return ggplot object
#' @export
#'
#' @examples
#' ngenes <- 1000
#' nsamples <- 16
#' edat <- matrix(rpois(ngenes*nsamples, lambda=5), nrow=ngenes)
#' rownames(edat) <- paste0("gene_", 1:ngenes)
#' colnames(edat) <- paste0("sample", 1:nsamples)
#' cl <- CoexList(counts = edat)
#' colData(cl) <- S4Vectors::DataFrame(row.names = colnames(edat),
#' group = c(rep("A", 8), rep("B", 8)))
#' cl <- normCounts(cl)
#' plotSampleTree(cl, plotColVar=TRUE, colVar="group")

plotSampleTree <- function(cl, plotColVar=FALSE, colVar=1, corMethod="spearman",
                           hclustMethod="average"){

    stopifnot("cl must be a CoexList object" = class(cl)[1] == "CoexList")
    stopifnot("plotColVar must be logical of length 1" =
                  class(plotColVar) == "logical" &
                  length(plotColVar) == 1)

    # Check colVar is valid
    pass <- NULL

    if( plotColVar == TRUE ){
        pass <- (
            is.numeric(colVar) &
            length(colVar) == 1 &
            colVar %in% 1:ncol(colData(cl))
            ) | (
                is.character(colVar) &
                    length(colVar) == 1 &
                    colVar %in% colnames(colData(cl))
            )

        stopifnot("colVar must be a numeric or character of length 1 that refers to a column of colData(cl)" =
                      pass == TRUE)
    }

    # Do the clustering
    cat("Calculating correlations...\n")
    dissimilarity <- 1 - stats::cor(cl@normCounts, method = corMethod)
    cat("Clustering...\n")
    distance <- as.dist(dissimilarity)
    hc <- hclust(distance, method = hclustMethod)
    gg <- suppressMessages(
        ggdendro::ggdendrogram(hc, rotate = FALSE) +
            ggplot2::scale_y_continuous(expand = c(0,0)) +
            ggplot2::theme(axis.text.y = ggplot2::element_blank(),
                           plot.margin = ggplot2::unit(c(10,10,0,10), "points")) +
            coord_cartesian(xlim = c(1, length(hc$labels)))

    )

    # Get the colData for colour bar

    if( plotColVar == TRUE ){

        x <- NULL
        lab <- NULL

        if (class(colVar) == "character"){
            x <- match(colVar, colnames(colData(cl)))
            lab <- colVar
        } else {
            x <- colVar
            lab <- colnames(colData(cl))[colVar]
        }

        get_col_data <- function(x){

            df <- data.frame(id = rownames(colData(cl)),
                             variable = colnames(colData(cl))[x],
                             value = colData(cl)[ ,x],
                             order = hc$order)

            # Set order of dendrogram
            df$id <- factor(df$id, levels = hc$labels[hc$order])

            return(df)
        }

        col_dat <- get_col_data(x)

        # Plot the colour bar
        gg2 <- ggplot2::ggplot(col_dat, aes(x = .data$id,
                                            y = .data$variable,
                                            fill=.data$value)) +
            ggplot2::geom_tile() +
            ggplot2::scale_x_discrete(expand = c(0,0)) +
            coord_cartesian(xlim = c(1, length(hc$labels))) +
            ggplot2::scale_y_discrete(expand=c(0,0)) +
            ggplot2::xlab("") +
            ggplot2::ylab("") +
            #labs(fill=) +
            ggplot2::theme(plot.margin = unit(c(0,10,10,10), "points"),
                           axis.ticks = ggplot2::element_blank(),
                           axis.text = ggplot2::element_blank(),
                           legend.position="bottom")

        # Assemble the plot
        gg <- cowplot::plot_grid(gg, gg2, axis = "l", align = "v",
                                 nrow = 2,
                                 rel_heights = c(0.4, 0.2))

    }

    return(gg)
}

```

