---
title: "coex"
subtitle: "An R package for straightforward co-expression analyses"
output: BiocStyle::html_document
#output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(coex)
library(SummarizedExperiment)
library(magrittr)
library(edgeR)
library(recount)
#library(easierData)
#library(ExperimentHub)
```

```{r}
file_url <- download_study('SRP041675', type = "rse-gene", download = FALSE)

# Load the data
load(file = url(file_url))
rse_gene
```


Load the recount data for SRP041675
```{r, eval=FALSE}
eh <- ExperimentHub::ExperimentHub()
list_easierData()
mariathasan_dat <- eh[["EH6677"]]
#mariathasan_dat %>% assay() %>% head()
#mariathasan_dat %>% colData()
#library(curatedAdipoRNA)
#data("eas   ")


```

Subset for data from Illumina HiSeq platform
```{r, eval=FALSE}
adipo_counts <- adipo_counts[ ,grepl(pattern = "Illumina HiSeq",
                                     x = adipo_counts@colData$instrument_model)]
```

Load the gene expression counts into a coexList object
```{r, eval=FALSE}
cl <- coexList(counts = assay(mariathasan_dat))
```

Add the gene and sample metadata to the coexList object
```{r, eval=FALSE}
colData(cl) <- colData(mariathasan_dat)
rowData(cl) <- rowData(mariathasan_dat)
```

Normalise the data for co-expression calculations using the CTF method and scale using `asinh()`.
https://github.com/krishnanlab/RNAseq_coexpression/blob/main/src/CTF_normalize.R
```{r, eval=FALSE}
cl <- normCounts(cl)
```

Filter the data
```{r, eval=FALSE}
checkFilter(cl, propGenes = 0.25)
cl <- applyFilter(cl, propGenes = 0.25)
```

Calculate soft power thresholds for WGCNA network
```{r, eval=FALSE}
cl <- calcSoftPower(cl)
```
Check what the estimated power is
```{r, eval=FALSE}
cl@powerEstimate
```

Plot soft power estimates
```{r, eval=FALSE}
plotSoftPower(cl)
```

Calc adjacency matrix
```{r, eval=FALSE}
cl <- calcAdjacency(cl, method = "wgcna", softPower = 10)
```

Calculate topological overlap and cluster
```{r, eval=FALSE}
cl <- calcTOM(cl, )
```

Cluster 
```{r, eval=FALSE}

```

Plot clustering
```{r, eval=FALSE}

```

Plot the modules for groups
```{r, eval=FALSE}

```


