---
title: "coex - An R package for straightforward co-expression analyses"
output: BiocStyle::html_document
#output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(coex)
library(recount)
library(magrittr)
library(edgeR)
```

Load the recount data for SRP041675
```{r}
project_info <- recount::download_study("SRP041675", download = FALSE)
load(file = url(project_info))
assay(rse_gene)[1:6, 1:3]
colData(rse_gene)$title %>% table()
```

Load the gene expression counts into a coexList object
```{r}
cl <- coexList(counts = assay(rse_gene))
```

Add the gene and sample metadata to the coexList object
```{r}
colData(cl) <- colData(rse_gene)
rowData(cl) <- rowData(rse_gene)
```

Normalise the data for co-expression calculations using the CTF method and scale using `asinh()`.
https://github.com/krishnanlab/RNAseq_coexpression/blob/main/src/CTF_normalize.R
```{r}
norm_factors <- calcNormFactors(assay(cl),
                                lib.size = colSums(assay(cl)),
                                method = "TMM")

CTF_norm <- sweep(assay(cl), 2, norm_factors, "/")
CTF_norm <- asinh(CTF_norm)

cl@normCounts <- CTF_norm
```

Filter the data
```{r}
checkFilter(cl, propGenes = 0.25)
cl <- applyFilter(cl, propGenes = 0.25)
```

Calculate soft power thresholds for WGCNA network
```{r}
#cl <- calcSoftPower(cl)
```

Check what the estimated power is
```{r}
#cl@powerEstimate
```

Plot soft power estimates
```{r}
#plotSoftPower(cl)
```

Calc adjacency matrix
```{r}

```

Calculate topological overlap and cluster
```{r}

```

Plot the clustering and modules
```{r}

```

Plot the modules for groups
```{r}

```


