---
title: "coex"
subtitle: "An R package for straightforward co-expression analyses"
output: BiocStyle::html_document
#output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(coex)
library(magrittr)
library(recount)
```

Load data from SRP041675 as a RangedSummarizedExperiment. This loads an object named `rse_gene`.
```{r}
file_url <- download_study('SRP041675', type = "rse-gene", download = FALSE)

# Load the data
load(file = url(file_url))
rse_gene
```

```{r, eval=FALSE}
load("~/Desktop/rse-gene.Rdata")
```

Load the gene expression counts into a coexList object
```{r}
cl <- coexList(counts = assay(rse_gene))
```

Add the gene and sample metadata to the coexList object
```{r}
colData(cl) <- colData(rse_gene)
rowData(cl) <- rowData(rse_gene)
```

Now we can remove the `rse_gene` object
```{r}
rm(rse_gene)
```

Normalise the data for co-expression calculations using the CTF method and scale using `asinh()`.
https://github.com/krishnanlab/RNAseq_coexpression/blob/main/src/CTF_normalize.R
```{r}
cl <- normCounts(cl)
```

Some sample clustering to identify outliers


Filter the data
```{r}
checkFilter(cl, propGenes = 0.1)
cl <- applyFilter(cl, propGenes = 0.1)
```

Calculate soft power thresholds for WGCNA network
```{r, eval=FALSE}
cl <- calcSoftPower(cl)
```

Check what the estimated power is
```{r, eval=FALSE}
cl@powerEstimate
```

Plot soft power estimates
```{r, eval=FALSE}
plotSoftPower(cl, eval=FALSE)
```

Calc adjacency matrix
```{r, eval=FALSE}
cl <- calcAdjacency(cl, method = "wgcna", softPower = 5)
```

Calculate topological overlap matrix (TOM)
```{r, eval=FALSE}
cl <- calcTOM(cl)
```
Cluster genes
```{r, eval=FALSE}
cl <- calcTree(cl)
```

Plot clustering
```{r, eval=FALSE}
plotTree(cl, pamStage=TRUE)
```


Plot the modules for groups
```{r, eval=FALSE}

```


